# Studies using refgenie

Here are some studies we have found that make use of refgenie in some way:

<li><b>Dozmorov et al. (2021). </b><i>Chromatin conformation capture (Hi-C) sequencing of patient-derived xenografts: analysis guidelines</i> 
<br><i>GigaScience</i>.  <span class="doi">DOI: <a href="http://dx.doi.org//10.1093/gigascience/giab022">/10.1093/gigascience/giab022</a></li>
<li><b>Leal-Calvo et al. (2021). </b><i>A new paradigm for leprosy diagnosis based on host gene expression</i> 
<br><i>bioRxiv</i>.  <span class="doi">DOI: <a href="http://dx.doi.org/10.1101/2021.07.30.454441">10.1101/2021.07.30.454441</a></li>
<li><b>Rauscher et al. (2021). </b><i>Patient-derived gene and protein expression signatures of NGLY1 deficiency</i> 
<br><i>bioRxiv</i>.  <span class="doi">DOI: <a href="http://dx.doi.org/10.1101/2021.07.28.453930">10.1101/2021.07.28.453930</a></li>
<li><b>Gudukbay et al. (2021). </b><i>GYAN: Accelerating Bioinformatics Tools in Galaxy with GPU-Aware Computation Mapping</i> 
<br> <span class="doi">DOI: <a href="http://dx.doi.org/10.1109/ipdpsw52791.2021.00037">10.1109/ipdpsw52791.2021.00037</a></li>
<li><b>Kupkova et al. (2021). </b><i>Histone H3 lysine 27 acetylation profile undergoes two global shifts in undernourished children and suggests one-carbon metabolite insufficiency</i> 
<br><i>medRxiv</i>.  <span class="doi">DOI: <a href="http://dx.doi.org/10.1101/2021.06.11.21258783">10.1101/2021.06.11.21258783</a></li>
<li><b>Toulmin et al. (2021). </b><i>Type II alveolar cell MHCII improves respiratory viral disease outcomes while exhibiting limited antigen presentation</i> 
<br><i>Nature Communications</i>.  <span class="doi">DOI: <a href="http://dx.doi.org/10.1038/s41467-021-23619-6">10.1038/s41467-021-23619-6</a></li>
<li><b>Gerber et al. (2021). </b><i>Streamlining differential exon and 3′ UTR usage with diffUTR</i> 
<br><i>BMC Bioinformatics</i>.  <span class="doi">DOI: <a href="http://dx.doi.org/10.1186/s12859-021-04114-7">10.1186/s12859-021-04114-7</a></li>
<li><b>Hasegawa et al. (2021). </b><i>Clonal inactivation of telomerase promotes accelerated stem cell differentiation</i> 
<br><i>bioRxiv</i>.  <span class="doi">DOI: <a href="http://dx.doi.org/10.1101/2021.04.28.441728">10.1101/2021.04.28.441728</a></li>
<li><b>Mölder et al. (2021). </b><i>Sustainable data analysis with Snakemake</i> 
<br><i>F1000Research</i>.  <span class="doi">DOI: <a href="http://dx.doi.org/10.12688/f1000research.29032.2">10.12688/f1000research.29032.2</a></li>
<li><b>Smith et al. (2021). </b><i>PEPPRO: quality control and processing of nascent RNA profiling data</i> 
<br><i>Genome Biology</i>.  <span class="doi">DOI: <a href="http://dx.doi.org/10.1186/s13059-021-02349-4">10.1186/s13059-021-02349-4</a></li>
<li><b>Smith et al. (2020). </b><i>PEPATAC: An optimized ATAC-seq pipeline with serial alignments</i> 
<br><i>bioRxiv</i>.  <span class="doi">DOI: <a href="http://dx.doi.org/10.1101/2020.10.21.347054">10.1101/2020.10.21.347054</a></li>
<li><b>VijayKrishna et al. (2020). </b><i>Expanding the Galaxy's reference data</i> 
<br><i>bioRxiv</i>.  <span class="doi">DOI: <a href="http://dx.doi.org/10.1101/2020.10.09.327114">10.1101/2020.10.09.327114</a></li>
<li><b>Zhou et al. (2020). </b><i>CATA: a comprehensive chromatin accessibility database for cancer</i> 
<br><i>bioRxiv</i>.  <span class="doi">DOI: <a href="http://dx.doi.org/10.1101/2020.05.16.099325">10.1101/2020.05.16.099325</a></li>